episoder.1
.TH EPISODER 1
.SH NAME
episoder \- TV show episode reminder.
.SH SYNOPSIS
.B episoder
[global options] [command] [command-specific options]
.SH COPYRIGHT
episoder is Copyright (C) 2004-2017 by Stefan Ott
.SH DESCRIPTION
episoder is a tool to tell you about new episodes of your favourite TV shows
.SH GLOBAL OPTIONS
.TP
.B \-h
Show help and quit.
.TP
.B \-c \fIFILE\fR
Use \fIFILE\fR for configuration values. By default ~/.episoder is used.
.TP
.B \-v, \-\-verbose
Enable verbose operation.
.TP
.B \-d, \-\-debug
Enable debug (very verbose) operation.
.TP
.B \-l \fIFILE\fR
Log to \fIFILE\fR instead of stdout.
.TP
.B \-V, \-\-version
Show program version and quit.
.SH COMMANDS
.SS Add Shows
.B episoder \fR[options] \fBadd\fR \fISHOW\fR
.PP
This adds a \fISHOW\fR to episoder's database. You can either specify a URL from epguides.com or a show ID from TheTVDB.com. Note that this won't fetch any episodes or meta data about the show, to do this you will need to \fBupdate\fR your database.
.SS Find Shows on TheTVDB.com
.B episoder \fR[options] \fBsearch \fI"search terms"\fR
.PP
Rather than visiting their web site you can use episoder to search TheTVDB.com.  This command will show you a list of all shows that match your \fIsearch terms\fR along with their IDs -- you will need to know these ids to \fBadd\fR shows to the database.
.SS List Shows
.B episoder \fR[options] \fBshows\fR
.PP
List all shows in episoder's database. The output contains information on the show's URL or TVDB ID, name, current state, the time of the last update and the number of episodes currently in the database.
.PP
The \fBshows\fR command knows the following extra options:
.TP
.B \-a, \-\-active
Only include shows that are currently running or suspended in the list.
.SS Remove Shows
.B episoder \fR[options] \fBremove \fISHOW\fR
.PP
This removes a \fISHOW\fR from episoder's database. Please specify the \fIshow id\fR as shown by the \fBshows\fR command.
.SS Disable Updates
.B episoder \fR[options] \fBdisable \fISHOW\fR
.PP
By default, any show that is added to episoder's database is included when the database is updated. If however you would like to disable updates for a certain show, this is how you do it.
.PP
Please specify the \fIshow id\fR as shown by the \fBshows \fRcommand.
.SS Enable Updates
.B episoder \fR[options] \fBenable \fISHOW\fR
.PP
Use this to re-enable updates for a show that has previously been \fIdisabled\fR. Please specify the \fIshow id\fR as shown by the \fBshows \fRcommand.
.SS Update the Database
.B episoder \fR[options] \fBupdate \fR[update-options]
.PP
Update the episoder database. This will fetch the list of known episodes for all shows that are currently in the database (unless they have been disabled).
.PP
The \fBupdate \fRcommand knows the following extra options:
.TP
.B \-d \fIDATE\fR
All episodes that are older than \fIDATE\fR will be removed from the database after the update. You can either specify an absolute date as \fIYYYY-MM-DD\fR or a relative date by passing the number of days back. The default value is 2, i.e.  all episodes more than two days old will be removed.
.TP
.B \-f, \-\-force
Update the database regardless of the time of the last update. By default episoder uses these update intervals:
.RS
.PP
* 2 days for shows that are currently running.
.PP
* 1 week for shows that are currently paused.
.PP
* 2 weeks for shows that have ended.
.RE
.TP
.B \-i, \-\-nodate
Don't remove any old episodes, regardless of their date.
.TP
.B \-s, \-\-show \fIID\fR
Only update the show with this \fIID\fR (as shown by the \fBshows \fRcommand), regardless of the last update time and whether or not updates are currently enabled for this show.
.TP
.B \-n, \-\-num \fINUM\fR
Limit the number of shows to be updated to \fINUM\fR. If you have a lot of shows in your database this option allows you to prevent load spikes, i.e. to distribute your updates more evenly by only updating a few of them at a time.
.TP
Note that the \fB-d\fP and \fB-i\fP options cannot be combined and that \fB-f\fP has no effect when using \fB-s\fP.
.SS List Episodes in the Database
.B episoder \fR[options] \fBlist \fR[list-options]
.PP
Show upcoming episodes. Before running this you should update the database to make sure that your list of episodes is up-to-date.
.PP
The \fBlist \fRcommand knows the following extra options:
.TP
.B \-C, \-\-nocolor
Don't use colors in the list of episodes. By default, episodes will be colored to indicate the air date relative to the current date:
.RS
.PP
* Episodes from the distant past are colored gray
.PP
* Yesterday's episodes are colored red
.PP
* Today's episodes are colored yellow
.PP
* Tomorrow's episodes are colored green
.PP
* Future episodes are colored cyan
.RE
.TP
.B \-d \fIDATE\fR
Only show episodes that air prior to \fIDATE\fR. You can either specify an absolute date as \fIYYYY-MM-DD\fR or a relative date by passing the number of days back. The default is to hide all episodes that aired before yesterday.
.TP
.B \-n \fIDAYS\fB, \-\-days \fIDAYS\fR
The number of days to show. Any episodes that are more than \fIDAYS\fR days in the future will be hidden. The default value is 2.

Note that this value is relative to the \fIDATE\fR specified with the \fB\-d \fRoption.
.TP
.B \-i, \-\-nodate
Show all episodes regardless of their date.
.TP
.B \-s \fISEARCH\fR
Search for \fISEARCH\fR in show and episode names.
.SS Send Notifications
.B episoder \fR[options] \fBnotify \fR[notify-options]
.PP
Send e-mail notifications about recently discovered episodes. You can run this command after \fBepisoder update \fRin order to get a friendly reminder about upcoming episodes to your INBOX.
.PP
The \fBnotify \fRcommand knows the following extra options:
.TP
.B \-d \fIDATE\fR
Only report episodes that air prior to \fIDATE\fR. You can either specify an absolute date as \fIYYYY-MM-DD\fR or a relative date by passing the number of days back. The default is TODO.
.TP
.B \-n \fIDAYS\fB, \-\-days \fIDAYS\fR
The number of days to include. Any episodes that are more than \fIDAYS\fR days in the future will be ignored. The default value is TODO (2?).
.TP
.B \-s \fIID\fB, \-\-show \fIID\fR
Only send notifications about episodes that belong to the show with the id \fIID\fR.
.TP
Note that you will need to configure your e-mail settings in the configuration file (see below for details).
.SH EXAMPLES
Find a show on TheTVDB.com:
.IP
.B episoder search dexter
.IP
The output will look like this:
.IP
.nf
ID      Name
------- --------------------
79349   Dexter
.fi

.TP
Add a show:
.IP
.B episoder add 79349

.TP
List all currently added shows:
.IP
.B
episoder shows
.IP
The output will look like this:
.IP
[   1] 79349
       Dexter, Running, Enabled
       Last update: 1970-01-01 00:00:00
       Episodes: 0

.TP
Fetch all episodes for Dexter regardless of the air date and enable debugging output:
.IP
.B episoder -d update -s 1 -i

.TP
List all episodes from October 2012:
.IP
.B episoder list -d 2012-10-01 -n 31
.SH CONFIG FILE
.LP
The configuration file contains default settings for episoder.
.TP
.B agent=foo
Set the user-agent string to be used when fetching data. This is only used for shows on epguides.com.
.TP
.B data=/path/to/file
Path to episoder's data file. If, instead of a file, you supply a database url (as expected by sqlalchemy, eg. 'mysql://localhost/episoder'), episoder is going to use that database for storage instead.
.TP
.B format=unquoted format string
This allows you to customize episoder's output. Available fields are:
.PP
        %airdate     The episode's airdate as YYYY-MM-DD
        %show        Name of the show
        %season      Current season
        %epnum       Episode's number in season
        %eptitle     Title of the episode
        %totalep     Episode's total number
        %prodnum     Production number
.br

If undefined, the default value of \fB%airdate %show %seasonx%epnum \fR(eg. "2005-07-29 Monk 4x04") is used.
.TP
.B dateformat=unquoted string describing the format
Here you can define the date format you'd like to be used for the output. To get a list of all possible fields, see
.BR date(1).

The default is %a, %b %d, %Y
.TP
.B tvdb_key=KEY
Use your own TVDB API \fIKEY\fR instead of episoder's built-in default key. This is recommended if you use episoder on a larger scale.

See http://thetvdb.com/?tab=apiregister for more details.
.TP
.B email_to=user@example.org
This is the e-mail address that will be used for notifications about new episodes. You will need to enter an address here to use the \fBnotify\fR feature.
.TP
.B email_server=localhost\fR, \fBemail_port=587
If you want to use an e-mail server other than localhost for sending notifications you can specify it here, along with the port number to connect to.
.TP
.B email_username=username\fR, \fBemail_password=password
If you need to authenticate against your mail server, this is where you can enter your user name and password.
.TP
.B email_tls=0
Set this to 1 to use TLS encryption to send your mail.
.SH CRON
.LP
You might want to have your episode database rebuilt on a regular basis (i.e. daily). The easiest way to achieve this is with a simple cron job:
.PP
    crontab \-l > crontab
    echo "40 5 * * * episoder update" >> crontab
    crontab crontab

.SH FILES
.B ~/.episoder
- default configuration file
.br
.B ~/.episodes
- default data file

.SH AUTHOR
This manual page was written by Stefan Ott


CHANGELOG
== v0.8.1 (UNRELEASED) ==
 * Handle failures to load TVDB API pages
 * Correctly update tv show timestamp when updating shows

== v0.8.0 (7 Jun 2017) ==
 * Switched from distutils to setuptools
 * We now use the new TVDB JSON API
 * Rewrote the epguides.com parser in pure Python
 * Rewrote large parts of the database code
 * Episoder should now work fully with Python 3
 * The update command can now be limited to a fixed number of shows
 * Updated man page
 * Thanks to Chris Spencer (chrisspen) for his contributions
 * Added option to only list active shows (chrisspen)
 * Added the number of episodes to the list of shows (chrisspen)
 * Added email notification feature via 'episoder notify' (chrisspen)

== v0.7.3 (7 Nov 2015) ==
 * Fixed a syntax error in the list of shows
 * Fixed season number parsing for epguides.com (thanks to jethrogb)

== v0.7.2 (26 Sep 2015) ==
 * Dropped unneeded dependency on BeautifulSoup
 * Fixed episode number parsing for epguides.com (thanks to jethrogb)
 * Improved python3 compatibility

== v0.7.1 (27 Apr 2014) ==
 * Accept unicode characters in prodnum fields

== v0.7.0 (14 Oct 2013) ==
 * Added support for TheTVDB.com
 * Removed support for tv.com
 * Changed command-line arguments. See the man page for details
 * Removed the ability to specify shows in the configuration file
 * Reorganized the command line options. Check the documentation for details
 * Updated the epguides.com parser to ignore trailer links in episode names
 * Fixed some more encoding issues in the console renderer
 * Changed license to GPLv3

== v0.6.5 (6 Oct 2010) ==
 * Updated the epguides.com parser to ignore links to trailers in episode names

== v0.6.4 (25 Jun 2010) ==
 * Filter javascript from tv.com pages because it confuses beautifulsoup 3.1

== v0.6.3 (23 Jun 2010) ==
 * The epguides parser now properly reads the season number for more shows

== v0.6.2 (08 Nov 2009) ==
 * Updated tv.com parser (show names need to be lower-case now!)
 * Added -S option to specify a show id for (forced) updates

== v0.6.1 (26 Aug 2009) ==
 * Episoder can now talk to any database supported by sqlalchemy (see README)
 * Tell awk to process utf8 (for epguides parser)
 * Catch (and properly log) some more HTTP exceptions

== v0.6.0 (8 Aug 2009) ==
 * Episoder now keeps a persistent database of shows and episodes
 * Added new options (-L, -a, -r, -E, -D) to manage the database
 * The parsers now know whether a show is still running or not
   * Active (currently running) shows are updated every two days
   * Inactive (ended) shows are updated every two weeks
   * The new -B command overrides those intervals and forces an update
 * Show names can no longer be overridden in the configuration file
 * Don't report an error when no sources are defined in the config file
 * The tv.com parser now also accepts urls containing '!', '+' or '%'
 * The -d option now also accepts the number of days instead of a fixed date
 * Documentation updated

== v0.5.3 (26 Jul 2009) ==
 * Added support for another epguides formatting style (used eg. for "Eureka")
 * Added option to suppress colors in console output
 * Added a filter to remove invalid characters from epguides content
 * Fixed some other parsing issues in the epguides parser
 * Improved compatibility with older python versions

== v0.5.2 (30 Jun 2009) ==
 * Reworked parts of the code to improve compatibility with older libraries
 * Fixed some encoding issues in the console renderer

== v0.5.1 (23 Jun 2009) ==
 * Thanks to Philip Douglass for his bug reports and patches
 * The epguides parser now properly closes all file handles and works with mawk
 * The tv.com parser would crash on missing production/season/episode numbers
 * The tv.com parser now accepts urls that contain a dash
 * Switched to using sqlalchemy to access the database
 * Properly implemented the user-agent configuration option
 * Removed old FAQ
 * Fixed some text encoding issues
 * Small fixes to the man page

== v0.5.0 (18 May 2009) ==
 * Rewrote the whole thing in python
 * Wrote a brand new tv.com parser
 * Small fixes in the epguides.com parser
 * Parsed data is now stored in an sqlite3 database

== v0.4.7 (24 Feb 2007) ==
 * Fixed a parsing issue introduced with 0.4.6

== v0.4.6 (24 Feb 2007) ==
 * Discovered yet another formatting possibility on epguides.com (used for
   instance for the show "Midsomer Murders") and adapted the parser accordingly
 * Slightly updated the man page

== v0.4.5 (23 Jan 2007) ==
 * Fixed some epguides parser issues

== v0.4.4 (19 Nov 2006) ==
 * Fixed the -i option when building the database (did the exact opposite of
   what it was supposed to do)

== v0.4.3 (24 Oct 2006) ==
 * We now actually honor the -i option when building the database
 * Added the possibility to override a show's name in the config file. This was needed for the new webisoder, a web-based episoder. Find out more about it at http://www.webisoder.net/
 * Oops, forgot to update episoder's version number in the last release...
 * Fixed a bug in the new epguides.com parser which would not drop episodes without a date

== v0.4.2 (21 Oct 2006) ==
 * Don't ignore the config file passed with -c anymore when building the db
 * Added parser for epguides.com (closes #38)

== v0.4.1 (20 Apr 2006) ==
 * Did some minor internal changes to get episoder to work on non-linux systems
 * Removed long options
 * Hotfix for the tv.com website changes

== v0.4.0 (3 Dec 2005) ==
 * Re-organized options: -vv is now -w, -rc is now -c (or --rc), -nd is gone (use --nodate), -sa is gone (use -s and --nodate instead), -i is gone (see below)
 * Search is now alway case-insensitive
 * If no action is specified, episoder now defaults to showlist
 * Removed the tvtome dummy plugin

== v0.3.0 (9 Aug 2005) ==
 * Majeed Goldrick joined the project
 * Moved parts of the website to sourceforge
 * Removed a 1024-episodes-limit in sortlist
 * Added search options
 * Switched to awk to remove old episodes (>600 times faster)

pyepisoder/episoder.py
# episoder, https://github.com/cockroach/episoder
#
# Copyright (C) 2004-2017 Stefan Ott. All rights reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import absolute_import

import logging
from datetime import date, timedelta

import sqlite3
from sqlalchemy import Table, MetaData, create_engine, or_, and_
from sqlalchemy.orm import create_session

from .database import Episode, Show, Meta


class Database(object):

	def __init__(self, path):

		self._path = path
		self.logger = logging.getLogger("Database")

		self.open()
		self._initdb()

	def __str__(self):

		return "Episoder Database at %s" % self._path

	def __repr__(self):

		return "Database(%s)" % self._path

	def _initdb(self):

		# Initialize the database if all tables are missing
		tables = [Show, Episode, Meta]
		tables = map(lambda x: x.__table__.exists, tables)
		found = [x for x in tables if x(bind=self.engine)]

		if len(found) < 1:
			Show.__table__.create(bind=self.engine)
			Episode.__table__.create(bind=self.engine)
			Meta.__table__.create(bind=self.engine)
			self.set_schema_version(4)

	def open(self):

		if self._path.find("://") > -1:
			self.engine = create_engine(self._path,
							convert_unicode=True)
		else:
			self.engine = create_engine("sqlite:///%s" % self._path)

		self.conn = self.engine.connect()
		self.metadata = MetaData()
		self.metadata.bind = self.engine
		self.session = create_session(bind=self.engine)
		self.session.begin()

	def close(self):

		self.session.commit()
		self.session.close()
		self.conn.close()
		self.engine.dispose()

	def set_schema_version(self, version):

		meta = Meta()
		meta.key = "schema"
		meta.value = "%d" % version

		self.session.merge(meta)
		self.session.flush()

	def get_schema_version(self):

		if not Meta.__table__.exists(bind=self.engine):
			return 1

		res = self.session.query(Meta).get("schema")
		if res:
			return int(res.value)

		return 0

	def clear(self):

		episodes = self.session.query(Episode).all()

		for episode in episodes:
			self.session.delete(episode)

		self.session.flush()

	def migrate(self):

		schema_version = self.get_schema_version()
		self.logger.debug("Found schema version %s", schema_version)

		if schema_version < 0:

			self.logger.debug("Automatic schema updates disabled")
			return

		if schema_version == 1:

			# Upgrades from version 1 are rather harsh, we
			# simply drop and re-create the tables
			self.logger.debug("Upgrading to schema version 2")

			table = Table("episodes", self.metadata, autoload=True)
			table.drop()

			table = Table("shows", self.metadata, autoload=True)
			table.drop()

			Show.__table__.create(bind=self.engine)
			Episode.__table__.create(bind=self.engine)
			Meta.__table__.create(bind=self.engine)

			schema_version = 4
			self.set_schema_version(schema_version)

		if schema_version == 2:

			# Add two new columns to the shows table
			self.logger.debug("Upgrading to schema version 3")

			# We can only do this with sqlite databases
			assert self.engine.driver == "pysqlite"

			self.close()

			upgrade = sqlite3.connect(self._path)
			upgrade.execute("ALTER TABLE shows "
					"ADD COLUMN enabled TYPE boolean")
			upgrade.execute("ALTER TABLE shows "
					"ADD COLUMN status TYPE integer")
			upgrade.close()

			self.open()
			schema_version = 3
			self.set_schema_version(schema_version)

		if schema_version == 3:

			# Add a new column to the episodes table
			self.logger.debug("Upgrading to schema version 4")

			# We can only do this with sqlite databases
			assert self.engine.driver == "pysqlite"

			self.close()

			upgrade = sqlite3.connect(self._path)
			upgrade.execute("ALTER TABLE episodes "
					"ADD COLUMN notified TYPE date")
			upgrade.close()

			self.open()
			schema_version = 4
			self.set_schema_version(schema_version)

	def get_expired_shows(self, today=date.today()):

		delta_running = timedelta(2)	# 2 days
		delta_suspended = timedelta(7)	# 1 week
		delta_ended = timedelta(14)	# 2 weeks

		shows = self.session.query(Show).filter(or_(
				and_(
					Show.enabled,
					Show.status == Show.RUNNING,
					Show.updated < today - delta_running
				),
				and_(
					Show.enabled,
					Show.status == Show.SUSPENDED,
					Show.updated < today - delta_suspended
				),
				and_(
					Show.enabled,
					Show.status == Show.ENDED,
					Show.updated < today - delta_ended
				)
		))

		return shows.all()

	def get_enabled_shows(self):

		shows = self.session.query(Show).filter(Show.enabled)
		return shows.all()

	def get_show_by_url(self, url):

		shows = self.session.query(Show).filter(Show.url == url)

		if shows.count() < 1:
			return None

		return shows.first()

	def get_show_by_id(self, show_id):

		return self.session.query(Show).get(show_id)

	def add_show(self, show):

		show = self.session.merge(show)
		self.session.flush()
		return show

	def remove_show(self, show_id):

		show = self.session.query(Show).get(show_id)

		if not show:
			self.logger.error("No such show")
			return

		episodes = self.session.query(Episode)

		for episode in episodes.filter(Episode.show_id == show.id):
			self.session.delete(episode)

		self.session.delete(show)
		self.session.flush()

	def get_shows(self):

		return self.session.query(Show).all()

	def add_episode(self, episode, show):

		episode.show_id = show.id
		self.session.merge(episode)
		self.session.flush()

	def get_episodes(self, basedate=date.today(), n_days=0):

		enddate = basedate + timedelta(n_days)

		return self.session.query(Episode).\
			filter(Episode.airdate >= basedate). \
			filter(Episode.airdate <= enddate). \
			order_by(Episode.airdate).all()

	def search(self, search):

		return self.session.query(Episode).\
			filter(or_( \
				Episode.title.like("%%%s%%" % search),
				Show.name.like("%%%s%%" % search))). \
			order_by(Episode.airdate).all()

	def commit(self):

		self.session.commit()
		self.session.begin()

	def rollback(self):

		self.session.rollback()
		self.session.begin()

	def remove_before(self, then, show=None):

		eps = self.session.query(Episode).filter(Episode.airdate < then)

		if show:
			eps = eps.filter(Episode.show == show)

		for episode in eps:
			self.session.delete(episode)

		self.commit()
